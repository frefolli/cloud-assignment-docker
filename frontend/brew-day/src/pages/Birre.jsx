import React, { Component } from "react";
import BeerView from "../components/BeerView";
import BeerEdit from "../components/BeerEdit";
import BeerDelete from "../components/BeerDelete";
import Modal from "../components/Modal";
import MButton from "../components/MButton";
import BodyThemeManager from '../components/BodyThemeManager'
import {BEER_LIST_ENDPOINT, BEERS_ENDPOINT, RECIPE_ENDPOINT  } from '../Protocol';
import Selector from "../components/Selector";
import BeerTable from "../components/BeerTable";
import JimTable from "../components/JimTable";


class Birre extends Component {
  constructor(props) {
    super(props);
    this.state = {
      beerIDs: [],
      beers: [],
      currentAction: "",
      selectedBeer: null,
      showModal: false,
      filterRecipe: "",
      filterName: "",
      beerIDsFiltered: [],
      recipes: [],
    };
  }

  triggerReload = () => {
    fetch(BEER_LIST_ENDPOINT)
      .then((response) => response.json())
      .then((data) => {
        const beerIDs = data;
        const beerIDsFiltered = data;
        const promises = beerIDs.map((id) => fetch(BEERS_ENDPOINT+`${id}`));
        Promise.all(promises).then((results) => {
          const beers = results.map((response) => response.json());
          Promise.all(beers)
            .then((updatedBeers) => {
              this.setState({
                beerIDs,
                beers: updatedBeers,
                beerIDsFiltered,
              });
              return updatedBeers
                .map((beer) => beer.recipeID)
                .filter((value, index, self) => (self.indexOf(value) === index) && value !== null);
            })
            .then((updatedBeers) => {
              Promise.all(
                updatedBeers.map((beer) => fetch(RECIPE_ENDPOINT+`${beer}`))
              )
                .then((responses) =>
                  Promise.all(responses.map((response) => response.json()))
                )
                .then((data) => {
                  this.setState({ recipes: data });
                });
            });
        });
      })
      .catch((error) => console.error(error));
  };

  componentDidMount() {
    this.triggerReload();
  }

  handleView = (item) => {
    this.setState({
      currentAction: "view",
      selectedBeer: item,
      showModal: true,
    });
  }

  handleEdit = (item) => {
    this.setState({
      currentAction: "edit",
      selectedBeer: item,
      showModal: true,
    });
  }

  handleDelete = (item) => {
    this.setState({
      currentAction: "delete",
      selectedBeer: item,
      showModal: true,
    });
  }

  setShowModal = (flag) => {
    this.setState({ showModal: flag });
  };

  handleDeleteConfirm = () => {
    const beerToDeleteID = this.state.selectedBeer.beerID;
     fetch(BEERS_ENDPOINT+`${beerToDeleteID}`, {
      method: "DELETE",
    }).then(() => {
      this.triggerReload();
      this.setShowModal(false);
    })
  };

  getCurrentComponent = () => {
    let selectedBeer = this.state.selectedBeer;
    let currentAction = this.state.currentAction;

    if (selectedBeer === null) {
      return null;
    }

    switch (currentAction) {
      case "view":
        return <BeerView beerID={selectedBeer.beerID} />;
      case "edit":
        return (
          <BeerEdit
            beerID={selectedBeer.beerID}
            onConfirm={this.triggerReload}
          />
        );
      case "delete":
        return (
          <BeerDelete
            beerID={selectedBeer.beerID}
            onConfirm={this.handleDeleteConfirm}
          />
        );
      default:
        return null;
    }
  }

  setFilterName = (event) => {
    let filterName = event.target.value;
    this.setState({ filterName: filterName });
  }

  setFilterRecipe = (event) => {
    let filterRecipe = event.target.value;
    this.setState({ filterRecipe: filterRecipe });
  }

  render() {
    const { beerIDsFiltered, beers } = this.state;

    const beerItems = beerIDsFiltered.map((item) => {
      const beer = beers.find((b) => b.beerID === item);
      return beer
    });

    return (
      <BodyThemeManager>
        <div>
          <JimTable>
            <table style={{width: "100%"}}>
              <thead>
                <tr>
                  <th width="20%">FILTRA PER NOME</th>
                  <th width="50%">
                    <input
                      value={this.state.filterName}
                      type="text"
                      style={{ width: "100%", textAlign: "center" }}
                      onChange={(event) => this.setFilterName(event)}
                    ></input>
                  </th>
                  <th width="30%">
                    <MButton text="Filtra" onClick={() => this.filterBeer()} />
                    <MButton text="Togli" onClick={() => this.removeFilter()} />
                  </th>
                </tr>
                <tr>
                  <th width="20%">FILTRA PER RICETTA</th>
                  <th width="50%">
                    <Selector
                      label="Recipe"
                      value={this.state.filterRecipe}
                      onChange={this.setFilterRecipe}
                      options={this.state.recipes.map((recipe) => { return {name: recipe.name, value: recipe.recipeID}; })}
                    />
                  </th>
                  <th width="30%">
                    <MButton text="Filtra" onClick={() => this.filterBeer()} />
                    <MButton text="Togli" onClick={() => this.removeFilter()} />
                  </th>
                </tr>
              </thead>
            </table>
          </JimTable>
          <BeerTable
            beers={beerItems}
            handleView={this.handleView}
            handleEdit={this.handleEdit}
            handleDelete={this.handleDelete}
          />
          <Modal
            showModal={this.state.showModal}
            setShowModal={this.setShowModal}
          >
            {this.getCurrentComponent()}
          </Modal>
        </div>
      </BodyThemeManager>
    );
  }

  filterBeer = () => {
    let url = "";
    if (this.state.filterName === "") {
      if (this.state.filterRecipe === "") {
        url = BEER_LIST_ENDPOINT;
      } else {
        url = BEER_LIST_ENDPOINT + `?recipeID=${this.state.filterRecipe}`;
      }
    } else {
      if (this.state.filterRecipe === "") {
        url = BEER_LIST_ENDPOINT + `?name=${this.state.filterName}`;
      } else {
        url = BEER_LIST_ENDPOINT + `?name=${this.state.filterName}&&recipeID=${this.state.filterRecipe}`;
      }
    }

    fetch(url)
      .then((response) => response.json())
      .then((beerIDsFiltered) => {
        this.setState({ beerIDsFiltered: beerIDsFiltered });
      });
  }

  removeFilter = () => {
    this.setState({
      beerIDsFiltered: this.state.beerIDs,
      filterName: "",
      filterRecipe: "",
    });
  }
}
export default Birre;
