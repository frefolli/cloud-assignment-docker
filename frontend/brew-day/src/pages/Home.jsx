import React, { Component } from "react";
import RecipeExecute from "../components/RecipeExecute";

class Home extends Component {
  constructor(props) {
    super(props);
    this.state = {
      advice: null
    };
  }
  componentDidMount() {
    fetch("/api/advice")
      .then((res) => res.json())
      .then((data) => this.setState({ advice: data }))
      .catch(() => {});
  }
  render() {
    return (
      <div>
        <center>
          <h1 className="advice-texts">Vuoi un consiglio?</h1>
          {(this.state.advice === null) ? 
            <h1>Prova ad inserire qualche ricetta</h1>
          : (
            <div>
              <h1>Ecco quale birra dovresti preparare</h1>
              <RecipeExecute recipeID={this.state.advice.recipeID} color="white"/>
              <h3 className="advice-texts">La massima quantità realizzabile è di {this.state.advice.quantity}</h3>
            </div>
          )}
        </center>
      </div>
    );
  }
}

export default Home;