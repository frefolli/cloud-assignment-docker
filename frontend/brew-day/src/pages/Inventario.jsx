import React, { Component } from "react";

class Inventario extends Component {
  constructor(props) {
    super(props);
    this.state = { inventory: [], isLoading: true};
  }

  componentDidMount() {
    fetch("/api/inventory")
      .then((response) => response.json())
      .then((data) => this.setState({ inventory: data, isLoading: false }, () => {
        return <p>Caricamento...</p>
      }));
  }

  handleDelete = (name) => {
    fetch(`/api/inventory/${name}`, {
      method: "DELETE"
    }).then(() => {
      this.setState({
        inventory: this.state.inventory.filter(item => item.name !== name)
      });
    });
  };

  render() {
    const { inventory, isLoading } = this.state;

    if (isLoading) {
      return <p>Caricamento...</p>;
    }

    const itemList = inventory.map((item) => {
      let imagePath = `../../icons/inventory-icons/${item.name}.png`;
      let defaultImage = "../../icons/inventory-icons/sconosciuto.png";
      return (
        <tr key={item.name}>
          <td>
            <img
              className="shoppingImage"
              src={imagePath}
              alt = ""
              onError={(e) => { e.target.onerror = null; e.target.src=defaultImage }}
            />
          </td>
          <td>{item.name}</td>
          <td>{item.quantity}</td>
          <td>
            <button onClick={() => this.handleDelete(item.name)}>
              Elimina ingrediente
            </button>
          </td>
        </tr>
      );
    });

    return (
      <div>
        <table className="myTable">
          <thead>
            <tr>
              <th width="25%">Immagine</th>
              <th width="25%">Nome</th>
              <th width="25%">Quantit√†</th>
              <th width="25%">Azioni</th>
            </tr>
          </thead>
          <tbody>{itemList}</tbody>
        </table>
      </div>
    );
  }
}
export default Inventario;
