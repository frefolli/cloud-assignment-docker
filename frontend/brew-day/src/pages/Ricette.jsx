import React, { Component } from 'react';
import Modal from "../components/Modal";
import RecipeView from "../components/RecipeView";
import RecipeEdit from "../components/RecipeEdit";
import RecipeDelete from "../components/RecipeDelete";
import RecipeExecute from "../components/RecipeExecute";
import MButton from '../components/MButton';
import { RECIPE_LIST_ENDPOINT, SETTINGS_ENDPOINT, SETTING_LIST_ENDPOINT } from '../Protocol';
import Selector from '../components/Selector';
import RecipeTable from '../components/RecipeTable';
import JimTable from '../components/JimTable';
import ThemeManager from '../components/ThemeManager';
export default class Ricette extends Component {
    constructor(props) {
        super(props);
        this.state = {recipes: [], currentAction: "", nextRecipeQuantity: "", nextRecipeID: "", selectedRecipe: null, showModal:false, newRecipeName: "", newRecipeDescription: "", filterName: "", recipesFiltered: []};
    }

    triggerReload = () => {
        fetch(RECIPE_LIST_ENDPOINT)
        .then(response => response.json())
        .then(recipeIDs => Promise.all(recipeIDs.map(recipeID => fetch(`/api/recipes/${recipeID}`))))
        .then(responses => Promise.all(responses.map(response => response.json())))
        .then(data => this.setState({recipes: data, recipesFiltered: data, newRecipeName: "", newRecipeDescription: ""}));
    }

     triggerReloadSettings = () => {
        fetch(SETTINGS_ENDPOINT + "nextRecipeID")
        .then(response => response.json())
        .then(data => this.setState({nextRecipeID: data.value}))
        .catch((error) => {this.postNextRecipeSetting("nextRecipeID", "");
                          this.setState({nextRecipeID: ""});
                    });
        fetch(SETTINGS_ENDPOINT + "nextRecipeQuantity")
        .then(response => response.json())
        .then(data => this.setState({nextRecipeQuantity: data.value}))
        .catch((error) => {this.postNextRecipeSetting("nextRecipeQuantity", "0");
                          this.setState({nextRecipeQuantity: "0"});
                    });     
    }

    componentDidMount() {
      this.triggerReload();
      this.triggerReloadSettings();
    }

    handleView = (item) => {
      this.setState({currentAction:"view", selectedRecipe:item, showModal:true})
    };

    handleEdit = (item) => {
      this.setState({currentAction:"edit", selectedRecipe:item, showModal:true})
    };    
    
    handleDelete = (item) => {
      this.setState({currentAction:"delete", selectedRecipe:item, showModal:true});
      if(item.recipeID === this.state.nextRecipeID){
        this.updateNextRecipeSetting("nextRecipeID", "");
        this.updateNextRecipeSetting("nextRecipeQuantity", "0");
      }

    };

    handleExecute = (item) => {
      this.setState({currentAction:"execute", selectedRecipe:item, showModal:true})
    };

    closeModal = () => this.setShowModal(false);
    closeModalAndReload = () => {this.closeModal(); this.triggerReload()};

    getCurrentComponent = () => {
      let selectedRecipe = this.state.selectedRecipe;
      let currentAction = this.state.currentAction;
      if (!selectedRecipe) return <div>Caricamento...</div>;
      switch (currentAction) {
        case "view":
          return <RecipeView recipeID={selectedRecipe.recipeID}/>;
        case "edit":
          return <RecipeEdit recipeID={selectedRecipe.recipeID} onConfirm={this.triggerReload}/>;
        case "delete":
          return <RecipeDelete recipeID={selectedRecipe.recipeID} onConfirm={this.closeModalAndReload}/>;
        case "execute":
          return <RecipeExecute recipeID={selectedRecipe.recipeID} onConfirm={this.closeModal}/>;
        default:
          return <div></div>;
      }
    }

    setNewRecipeName = (event) => {
      let newRecipeName = event.target.value;
      this.setState({newRecipeName: newRecipeName});
    }

    setFilterName = (event) => {
      let filterName = event.target.value;
      this.setState({filterName: filterName});
    }
  
    setNewRecipeDescription = (event) => {
      let newRecipeDescription = event.target.value;
      this.setState({newRecipeDescription: newRecipeDescription});
    }

    setNewNextRecipeID = (event) => {
      let newNextRecipeID = event.target.value;
      this.setState({nextRecipeID: newNextRecipeID});
      this.updateNextRecipeSetting("nextRecipeID", newNextRecipeID)
    }

    setNewNextRecipeQuantity = (event) => {
      let newNextRecipeQuantity = event.target.value;
      this.setState({nextRecipeQuantity: newNextRecipeQuantity});
    }

    setShowModal = (flag) => {
      if (!flag)
        this.setState({currentAction:""})
      this.setState({showModal:flag})
    }
    
    removeFilter = () => {
      this.setState({recipesFiltered: this.state.recipes, filterName: ""})
    }

    render() {
        const {recipesFiltered, isLoading} = this.state;
        
        if (isLoading) {
            return <p>Caricamento...</p>;
        }
        
        return (
          <ThemeManager>
            <div>
              <JimTable>
                <table style={{width: "100%"}}>
                    <tbody>
                      <tr>
                        <td width="30%">FILTRA PER NOME</td>
                        <td width="50%">
                          <input
                            value={this.state.filterName} type="text"
                            style={{width: "100%", textAlign:"center"}}
                            onChange={ (event) => this.setFilterName(event)}
                          />
                        </td>
                        <td width="20%">
                          <MButton text="Filtra" onClick={() => this.filterRecipe()} />
                          <MButton text="Togli" onClick={() => this.removeFilter()} />
                        </td>
                      </tr>
                      <tr>
                        <td>Seleziona la prossima ricetta</td>
                        <td style={{display:"flex", flexFlow: "column nowrap", justifyContent:"space-around"}}>
                          <Selector
                            label="Recipe"
                            value={this.state.nextRecipeID}
                            onChange={this.setNewNextRecipeID}
                            options={this.state.recipes.map((recipe) => { return {name: recipe.name, value: recipe.recipeID}; })}
                          />
                          <input
                            style={{width: "100%", textAlign:"center"}}
                            value={this.state.nextRecipeQuantity} type="text"
                            onChange={ (event) => this.setNewNextRecipeQuantity(event)}
                          />
                        </td>
                        <td><MButton text="Programma" onClick={() => this.updateNextRecipeSetting("nextRecipeQuantity", this.state.nextRecipeQuantity)}/></td>
                      </tr>
                    </tbody>
                </table>
              </JimTable>

              <RecipeTable
                recipes={recipesFiltered}
                handleView={this.handleView}
                handleEdit={this.handleEdit}
                handleDelete={this.handleDelete}
                handleExecute={this.handleExecute}
                newRecipeName={this.state.newRecipeName}
                newRecipeDescription={this.state.newRecipeDescription}
                setNewRecipeName={this.setNewRecipeName}
                setNewRecipeDescription={this.setNewRecipeDescription}
                addRecipe={this.addRecipe}
                />
              <Modal showModal={this.state.showModal} setShowModal={this.setShowModal}>
                {this.getCurrentComponent()}
              </Modal>
            </div>
          </ThemeManager>
        );
    }

    addRecipe = () => {
      fetch(RECIPE_LIST_ENDPOINT, {
        method: 'POST',
        headers: {
            'Accept': 'application/json',
            'Content-Type': 'application/json'
        },
        body: JSON.stringify({name: this.state.newRecipeName, description: this.state.newRecipeDescription})
      })
      .then(() => this.triggerReload());
    }

    filterRecipe = () => {
      fetch(RECIPE_LIST_ENDPOINT + `?name=${this.state.filterName}`)
        .then(response => response.json())
        .then(recipesIDsFiltered => {
          let recipeFiltered = this.state.recipes.filter(recipe => recipesIDsFiltered.includes(recipe.recipeID));
          this.setState({recipesFiltered: recipeFiltered});
        })
    }

    updateNextRecipeSetting = (settingID, value) => {
      fetch(SETTINGS_ENDPOINT + `${settingID}`, {
          method: 'PUT',
          headers: {
              'Accept': 'application/json',
              'Content-Type': 'application/json'
          },
          body: JSON.stringify({value: value})
      })
    }

    postNextRecipeSetting = (settingID, value) => {
      fetch(SETTING_LIST_ENDPOINT, {
          method: 'POST',
          headers: {
              'Accept': 'application/json',
              'Content-Type': 'application/json'
          },
          body: JSON.stringify({settingID: settingID, value: value})
      });
    }
}