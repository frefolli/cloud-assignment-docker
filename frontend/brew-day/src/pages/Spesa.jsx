import React, { Component } from "react";
import { ThemeProvider } from "@mui/material";
import theme from "../theme/theme";
import MButton from "../components/MButton";

class Spesa extends Component {
  constructor(props) {
    super(props);
    this.state = {
      added: 0,
      ingredients: [
        {
          ingredientName: "",
          ingredientQuantity: "",
        },
      ],
    };
    this.handleSubmit = this.handleSubmit.bind(this);
    this.handleAddIngredient = this.handleAddIngredient.bind(this);
  }

  handleSubmit() {
    const _ingredients = [];
    for (const ingredient of this.state.ingredients) {
      const { ingredientName, ingredientQuantity } = ingredient;
      if (
        ingredientName &&
        ingredientQuantity &&
        this.state.ingredients.indexOf(ingredient) < this.state.added
      ) {
        _ingredients.push({
          name: ingredientName,
          quantity: ingredientQuantity,
        });
      }
    }
    if (_ingredients.length !== 0) {
      fetch("api/shopping/test", {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify(_ingredients),
      }).then(() => {
        this.setState({
          ingredients: [{ ingredientName: "", ingredientQuantity: "" }],
        });
      });
    }
  }
  handleDeleteIngredient(index) {
    this.state.added -= 1;
    this.setState({
      ingredients: this.state.ingredients.filter((p, i) => i !== index),
    });
  }

  handleAddIngredient() {
    const { ingredientName: newIngName, ingredientQuantity: newIngQuantity } =
      this.state.ingredients[this.state.ingredients.length - 1];
    const ingredients = this.state.ingredients;
    ingredients[ingredients.length - 1] = {
      ingredientName: newIngName,
      ingredientQuantity: newIngQuantity,
    };
    if (newIngName !== '' && newIngQuantity !== '') {
      ingredients.push({ ingredientName: "", ingredientQuantity: "" });
      this.setState({ ingredients });
    }
  }

  handleIngredientChange(field, index, value) {
    const ingredients = this.state.ingredients;
    ingredients[index][field] = value;
    this.setState({ ingredients });
  }

  render() {
    this.state.added = 0;
    return (
      <ThemeProvider theme={theme}>
        <table className="myTable">
          <thead>
            <tr>
              <th>Nome ingrediente</th>
              <th>Quantit√†</th>
              <th>Azione</th>
            </tr>
          </thead>
          <tbody>
            {this.state.ingredients.map((ingredient, index) => {
              let firstColumn = (
                <input
                  id="shoppingInputName"
                  data-testid="shopping-name"
                  type="text"
                  value={ingredient.ingredientName}
                  onChange={(e) =>
                    this.handleIngredientChange(
                      "ingredientName",
                      index,
                      e.target.value
                    )
                  }
                />
              );
              let button = (
                <MButton text="Aggiungi" onClick={this.handleAddIngredient} />
              );
              if (index !== this.state.ingredients.length - 1) {
                this.state.added += 1;
                button = (
                  <MButton
                    text="Elimina"
                    onClick={() => this.handleDeleteIngredient(index)}
                  />
                );
                firstColumn = <p>{ingredient.ingredientName}</p>;
              }
              return (
                <tr key={(this.state.ingredients.length - 1) === index ? 'default' : ingredient.ingredientName}>
                  <td>{firstColumn}</td>
                  <td>
                    <input
                      type="text"
                      data-testid="shopping-quantity"
                      value={ingredient.ingredientQuantity}
                      onChange={(e) =>
                        this.handleIngredientChange(
                          "ingredientQuantity",
                          index,
                          e.target.value
                        )
                      }
                    />
                  </td>
                  <td>{button}</td>
                </tr>
              );
            })}
          </tbody>
        </table>
        <div id="shoppingButtonContainer">
          <MButton text="Conferma" onClick={this.handleSubmit} />
        </div>
      </ThemeProvider>
    );
  }
}

export default Spesa;
