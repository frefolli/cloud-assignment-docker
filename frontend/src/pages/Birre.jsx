import React, { Component } from "react";
import BeerView from "../components/BeerView";
import BeerEdit from "../components/BeerEdit";
import BeerDelete from "../components/BeerDelete";
import Modal from "../components/Modal";
import MButton from "../components/MButton";
import BodyThemeManager from '../components/BodyThemeManager'
import {BEER_LIST_ENDPOINT, BEERS_ENDPOINT, RECIPE_ENDPOINT, FAKE_NOTIFIER  } from '../utils/Protocol';
import Selector from "../components/Selector";
import BeerTable from "../components/BeerTable";
import JimTable from "../components/JimTable";
import { TextField } from "@mui/material";
import JimFlex from "../components/JimFlex";
import JimGrid from "../components/JimGrid";


class Birre extends Component {
  constructor(props) {
    super(props);
    this.state = {
      beerIDs: [],
      beers: [],
      currentAction: "",
      selectedBeer: null,
      showModal: false,
      filterRecipe: "",
      filterName: "",
      beerIDsFiltered: [],
      recipes: [],
    };
    this.notifier = this.props.notifier || FAKE_NOTIFIER;
  }

  triggerReload = () => {
    fetch(BEER_LIST_ENDPOINT)
      .then((response) => response.json())
      .then((data) => {
        const beerIDs = data;
        const beerIDsFiltered = data;
        const promises = beerIDs.map((id) => fetch(BEERS_ENDPOINT+`${id}`));
        Promise.all(promises).then((results) => {
          const beers = results.map((response) => response.json());
          Promise.all(beers)
            .then((updatedBeers) => {
              this.setState({
                beerIDs,
                beers: updatedBeers,
                beerIDsFiltered,
              });
              return updatedBeers
                .map((beer) => beer.recipeID)
                .filter((value, index, self) => (self.indexOf(value) === index) && value !== null);
            })
            .then((updatedBeers) => {
              Promise.all(
                updatedBeers.map((beer) => fetch(RECIPE_ENDPOINT+`${beer}`))
              )
                .then((responses) =>
                  Promise.all(responses.map((response) => response.json()))
                )
                .then((data) => {
                  this.setState({ recipes: data });
                });
            });
        });
      })
      .catch(this.notifier.connectionError);
  };

  componentDidMount() {
    this.triggerReload();
  }

  handleView = (item) => {
    this.setShowModal(true);
    this.setState({
      currentAction: "view",
      selectedBeer: item
    });
  }

  handleEdit = (item) => {
    this.setShowModal(true);
    this.setState({
      currentAction: "edit",
      selectedBeer: item
    });
  }

  handleDelete = (item) => {
    this.setShowModal(true);
    this.setState({
      currentAction: "delete",
      selectedBeer: item
    });
  }

  setShowModal = (flag) => {
    if (!flag)
      this.setState({ currentAction: "" })
    this.setState({ showModal: flag });
  };

  handleDeleteConfirm = () => {
    const beerToDeleteID = this.state.selectedBeer.beerID;
     fetch(BEERS_ENDPOINT+`${beerToDeleteID}`, {
      method: "DELETE",
    })
    .then(this.notifier.onRequestError("impossibile eliminare la birra"))
    .then(this.notifier.onRequestSuccess("birra eliminata correttamente"))
    .then(() => {
      this.triggerReload();
      this.setShowModal(false);
    })
  };

  getCurrentComponent = () => {
    let selectedBeer = this.state.selectedBeer;
    let currentAction = this.state.currentAction;

    if (selectedBeer === null) {
      return null;
    }

    switch (currentAction) {
      case "view":
        return (
        <BeerView
          notifier={this.notifier}
          beerID={selectedBeer.beerID}
        />);
      case "edit":
        return (
          <BeerEdit
            notifier={this.notifier}
            beerID={selectedBeer.beerID}
            onConfirm={this.triggerReload}
          />
        );
      case "delete":
        return (
          <BeerDelete
            beerID={selectedBeer.beerID}
            onConfirm={this.handleDeleteConfirm}
          />
        );
      default:
        return null;
    }
  }

  setFilterName = (event) => {
    let filterName = event.target.value;
    this.setState({ filterName: filterName });
  }

  setFilterRecipe = (event) => {
    let filterRecipe = event.target.value;
    this.setState({ filterRecipe: filterRecipe });
  }

  render() {
    const { beerIDsFiltered, beers } = this.state;

    const beerItems = beerIDsFiltered.map((item) => {
      const beer = beers.find((b) => b.beerID === item);
      return beer
    });

    return (
      <BodyThemeManager>
        <div>
          <JimFlex>
            <JimTable>
                <p style={{textAlign:"center"}}>FILTRA PER NOME</p>
                <JimGrid>
                  <TextField
                    label="Name"
                    value={this.state.filterName}
                    type="text"
                    style={{ width: "100%", textAlign: "center" }}
                    onChange={(event) => this.setFilterName(event)}
                  />
                </JimGrid>
                <JimGrid>
                  <MButton text="Filtra" onClick={() => this.filterBeer()} />
                </JimGrid>
                <JimGrid>
                  <MButton text="Togli" onClick={() => this.removeFilter()} />
                </JimGrid>
              </JimTable>
              <JimTable>
                <p style={{textAlign:"center"}}>FILTRA PER RICETTA</p>
                <JimGrid>
                  <Selector optional
                    label="Recipe"
                    value={this.state.filterRecipe}
                    onChange={this.setFilterRecipe}
                    options={this.state.recipes.map((recipe) => { return {name: recipe.name, value: recipe.recipeID}; })}
                  />
                </JimGrid>
                <JimGrid>
                  <MButton text="Filtra" onClick={() => this.filterBeer()} />
                </JimGrid>
                <JimGrid>
                  <MButton text="Togli" onClick={() => this.removeFilter()} />
                </JimGrid>
              </JimTable>
          </JimFlex>
          
          <BeerTable
            beers={beerItems}
            handleView={this.handleView}
            handleEdit={this.handleEdit}
            handleDelete={this.handleDelete}
          />
          <Modal
            showModal={this.state.showModal}
            setShowModal={this.setShowModal}
          >
            {this.getCurrentComponent()}
          </Modal>
        </div>
      </BodyThemeManager>
    );
  }

  filterBeer = () => {
    let url = BEER_LIST_ENDPOINT+"?";
    if (this.state.filterName !== "")
      url += `&&name=${this.state.filterName}`;
    if (this.state.filterRecipe !== "")
      url += `&&recipeID=${this.state.filterRecipe}`;

    fetch(url)
    .then((response) => response.json())
    .then((beerIDsFiltered) => {
      this.setState({ beerIDsFiltered: beerIDsFiltered });
    })
    .catch(this.notifier.connectionError)
  }

  removeFilter = () => {
    this.setState({
      beerIDsFiltered: this.state.beerIDs,
      filterName: "",
      filterRecipe: "",
    });
  }
}
export default Birre;