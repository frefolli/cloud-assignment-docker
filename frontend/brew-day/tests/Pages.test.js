import React from "react";
import { render } from '@testing-library/react';
import Birre from '../src/pages/Birre';
import Home from '../src/pages/Home';
import Inventario from '../src/pages/Inventario';
import Ricette from '../src/pages/Ricette';
import Spesa from '../src/pages/Spesa';

describe('Birre component', () => {
    test('should render correctly', () => {
      const { container } = render(<Birre />);
      expect(container.firstChild).toMatchSnapshot();
    });
  });

describe('Home component', () => {
    test('should render correctly', () => {
      const { container } = render(<Home />);
      expect(container.firstChild).toMatchSnapshot();
    });
});

describe('Inventario component', () => {
    test('should render correctly', () => {
      const { container } = render(<Inventario />);
      expect(container.firstChild).toMatchSnapshot();
    });
});

describe('Ricette component', () => {
    test('should render correctly', () => {
      const { container } = render(<Ricette />);
      expect(container.firstChild).toMatchSnapshot();
    });
});

describe('Spesa component', () => {
    test('should render correctly', () => {
      const { container } = render(<Spesa />);
      expect(container.firstChild).toMatchSnapshot();
    });
});