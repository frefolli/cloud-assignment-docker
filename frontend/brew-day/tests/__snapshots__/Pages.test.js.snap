// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Birre component should render correctly 1`] = `<div />`;

exports[`Home component should render correctly 1`] = `<div />`;

exports[`Inventario component should render correctly 1`] = `
<div>
  <table
    class="myTable"
  >
    <tr>
      <td>
        <div>
          <img
            class="inventoryImage"
            src="../../icons/inventory-icons/acqua.png"
          />
          <p
            class="inventoryText"
          >
            Acqua
          </p>
          <p
            class="inventoryText"
          >
            Disponibilità: 
            0
          </p>
        </div>
      </td>
      <td>
        <div>
          <img
            class="inventoryImage"
            src="../../icons/inventory-icons/additivi.png"
          />
          <p
            class="inventoryText"
          >
            Additivi
          </p>
          <p
            class="inventoryText"
          >
            Disponibilità: 
            0
          </p>
        </div>
      </td>
      <td>
        <div>
          <img
            class="inventoryImage"
            src="../../icons/inventory-icons/lievito.png"
          />
          <p
            class="inventoryText"
          >
            Lievito
          </p>
          <p
            class="inventoryText"
          >
            Disponibilità: 
            0
          </p>
        </div>
      </td>
    </tr>
    <tr>
      <td>
        <div>
          <img
            class="inventoryImage"
            src="../../icons/inventory-icons/luppoli.png"
          />
          <p
            class="inventoryText"
          >
            Luppoli
          </p>
          <p
            class="inventoryText"
          >
            Disponibilità: 
            0
          </p>
        </div>
      </td>
      <td>
        <div>
          <img
            class="inventoryImage"
            src="../../icons/inventory-icons/malto.png"
          />
          <p
            class="inventoryText"
          >
            Malto
          </p>
          <p
            class="inventoryText"
          >
            Disponibilità: 
            0
          </p>
        </div>
      </td>
      <td>
        <div>
          <img
            class="inventoryImage"
            src="../../icons/inventory-icons/zuccheri.png"
          />
          <p
            class="inventoryText"
          >
            Zuccheri
          </p>
          <p
            class="inventoryText"
          >
            Disponibilità: 
            0
          </p>
        </div>
      </td>
    </tr>
  </table>
</div>
`;

exports[`Ricette component should render correctly 1`] = `<div />`;

exports[`Spesa component should render correctly 1`] = `
<div>
  <table
    class="myTable"
  >
    <tr>
      <th>
        Foto prodotto
      </th>
      <th>
        Nome - Descrizione
      </th>
      <th>
        Quantità
      </th>
    </tr>
    <tr>
      <td>
        <img
          class="shoppingImage"
          src="../../icons/inventory-icons/acqua.png"
        />
      </td>
      <td>
        Acqua
      </td>
      <td>
        <div>
          <input
            data-testid="quantity-input"
            id="shoppingItemId"
            name="shoppingItemId"
            value=""
          />
        </div>
      </td>
    </tr>
    <tr>
      <td>
        <img
          class="shoppingImage"
          src="../../icons/inventory-icons/additivi.png"
        />
      </td>
      <td>
        Additivi
      </td>
      <td>
        <div>
          <input
            data-testid="quantity-input"
            id="shoppingItemId"
            name="shoppingItemId"
            value=""
          />
        </div>
      </td>
    </tr>
    <tr>
      <td>
        <img
          class="shoppingImage"
          src="../../icons/inventory-icons/lievito.png"
        />
      </td>
      <td>
        Lievito
      </td>
      <td>
        <div>
          <input
            data-testid="quantity-input"
            id="shoppingItemId"
            name="shoppingItemId"
            value=""
          />
        </div>
      </td>
    </tr>
    <tr>
      <td>
        <img
          class="shoppingImage"
          src="../../icons/inventory-icons/luppoli.png"
        />
      </td>
      <td>
        Luppoli
      </td>
      <td>
        <div>
          <input
            data-testid="quantity-input"
            id="shoppingItemId"
            name="shoppingItemId"
            value=""
          />
        </div>
      </td>
    </tr>
    <tr>
      <td>
        <img
          class="shoppingImage"
          src="../../icons/inventory-icons/malto.png"
        />
      </td>
      <td>
        Malto
      </td>
      <td>
        <div>
          <input
            data-testid="quantity-input"
            id="shoppingItemId"
            name="shoppingItemId"
            value=""
          />
        </div>
      </td>
    </tr>
    <tr>
      <td>
        <img
          class="shoppingImage"
          src="../../icons/inventory-icons/zuccheri.png"
        />
      </td>
      <td>
        Zuccheri
      </td>
      <td>
        <div>
          <input
            data-testid="quantity-input"
            id="shoppingItemId"
            name="shoppingItemId"
            value=""
          />
        </div>
      </td>
    </tr>
  </table>
  <button
    id="shoppingBuyButton"
  >
    Compra
  </button>
</div>
`;
